src/Crypto/RNCryptor/V3/Decrypt.hs

```diff
@@ -135,9 +148,13 @@ decryptStream :: ByteString
 decryptStream userKey inS outS = do
   rawHdr <- S.readExactly 34 inS
   let hdr = parseHeader rawHdr
-  let ctx = newRNCryptorContext userKey hdr
-  processStream ctx inS outS decryptBlock finaliseDecryption
+      ctx = newRNCryptorContext userKey hdr
+      ctx' = ctx { ctxHMACCtx = update (ctxHMACCtx ctx) rawHdr }
+  processStream ctx' inS outS decryptBlock finaliseDecryption
   where
     finaliseDecryption lastBlock ctx = do
-      let (rest, _) = B.splitAt (B.length lastBlock - 32) lastBlock --strip the hmac
-      S.write (Just $ removePaddingSymbols (snd $ decryptBlock ctx rest)) outS
+      let (cipherText, msgHMAC) = B.splitAt (B.length lastBlock - 32) lastBlock
+          (ctx', clearText)     = decryptBlock ctx cipherText
+          hmac = convert $ finalize (ctxHMACCtx ctx')
+      when (not (consistentTimeEqual msgHMAC hmac)) (fail "failed decrypt, invalid HMAC")
```

Comment:
You can use `unless` instead of `when $ not ..`, they are semantically equivalent!